File: csharp/NeoC.Core/Lexer.cs
using System.Text;

namespace NeoC.Core;

// Lexer como ref struct para estado mutable eficiente (sin clases de dominio).
public ref struct Lexer
{
    private readonly string _source;
    private int _i;
    private int _line;
    private int _col;

    public Lexer(string source)
    {
        _source = source;
        _i = 0;
        _line = 1;
        _col = 1;
    }

    public List<Token> Tokenize()
    {
        var tokens = new List<Token>();
        while (true)
        {
            SkipWhitespaceAndComments();
            var line = _line;
            var col = _col;
            var ch = Peek();
            if (ch is null)
            {
                tokens.Add(new(TokenKind.Eof, string.Empty, line, col));
                break;
            }

            Token token = ch switch
            {
                '(' => Make(TokenKind.LParen, "(", line, col, 1),
                ')' => Make(TokenKind.RParen, ")", line, col, 1),
                '{' => Make(TokenKind.LBrace, "{", line, col, 1),
                '}' => Make(TokenKind.RBrace, "}", line, col, 1),
                ',' => Make(TokenKind.Comma, ",", line, col, 1),
                ';' => Make(TokenKind.Semicolon, ";", line, col, 1),
                '+' => Make(TokenKind.Plus, "+", line, col, 1),
                '*' => Make(TokenKind.Star, "*", line, col, 1),
                '/' => Make(TokenKind.Slash, "/", line, col, 1),
                '=' => Make(TokenKind.Equal, "=", line, col, 1),
                '-' when Peek(1) == '>' => Make(TokenKind.Arrow, "->", line, col, 2),
                '-' => Make(TokenKind.Minus, "-", line, col, 1),
                '"' => ReadString(),
                _ when char.IsAsciiDigit(ch.Value) => ReadInt(),
                _ when char.IsAsciiLetter(ch.Value) || ch == '_' => ReadIdentOrKeyword(),
                _ => throw new Exception($"Carácter no soportado '{ch}' en {line}:{col}")
            };

            tokens.Add(token);
        }

        return tokens;
    }

    private void SkipWhitespaceAndComments()
    {
        while (true)
        {
            while (Peek() is char ch && char.IsWhiteSpace(ch)) { _ = Advance(); }

            if (Peek() == '/' && Peek(1) == '/')
            {
                while (Peek() is char c && c != '\n') { _ = Advance(); }
                continue;
            }

            if (Peek() == '/' && Peek(1) == '*')
            {
                _ = Advance();
                _ = Advance();
                while (Peek() is char)
                {
                    if (Peek() == '*' && Peek(1) == '/')
                    {
                        _ = Advance();
                        _ = Advance();
                        break;
                    }
                    _ = Advance();
                }
                continue;
            }

            break;
        }
    }

    private Token ReadIdentOrKeyword()
    {
        var line = _line;
        var col = _col;
        var sb = new StringBuilder();

        while (Peek() is char ch && (char.IsAsciiLetterOrDigit(ch) || ch == '_'))
        {
            sb.Append(Advance());
        }

        var text = sb.ToString();
        var kind = text switch
        {
            "fn" => TokenKind.Fn,
            "struct" => TokenKind.Struct,
            "extern" => TokenKind.Extern,
            "import" => TokenKind.Import,
            "alias" => TokenKind.Alias,
            "return" => TokenKind.Return,
            "let" => TokenKind.Let,
            _ => TokenKind.Ident
        };

        return new(kind, text, line, col);
    }

    private Token ReadInt()
    {
        var line = _line;
        var col = _col;
        var sb = new StringBuilder();

        while (Peek() is char ch && char.IsAsciiDigit(ch))
        {
            sb.Append(Advance());
        }

        return new(TokenKind.IntLit, sb.ToString(), line, col);
    }

    private Token ReadString()
    {
        var line = _line;
        var col = _col;
        _ = Advance();

        var sb = new StringBuilder();
        while (Peek() is char ch)
        {
            if (ch == '"')
            {
                _ = Advance();
                return new(TokenKind.StringLit, sb.ToString(), line, col);
            }
            sb.Append(Advance());
        }

        throw new Exception($"String sin cerrar en {line}:{col}");
    }

    private Token Make(TokenKind kind, string lexeme, int line, int col, int len)
    {
        for (var k = 0; k < len; k++) { _ = Advance(); }
        return new(kind, lexeme, line, col);
    }

    private char? Peek(int look = 0) => _i + look < _source.Length ? _source[_i + look] : null;

    private char Advance()
    {
        var ch = _source[_i++];
        if (ch == '\n')
        {
            _line++;
            _col = 1;
        }
        else
        {
            _col++;
        }
        return ch;
    }
}

File: src/lexer.rs
use crate::token::{Token, TokenKind};

// Lexer simple con soporte de comentarios tipo C/C++.
pub struct Lexer<'a> {
    src: &'a [u8],
    pos: usize,
    line: usize,
    col: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(src: &'a str) -> Self {
        Self {
            src: src.as_bytes(),
            pos: 0,
            line: 1,
            col: 1,
        }
    }

    fn peek(&self) -> Option<u8> {
        self.src.get(self.pos).copied()
    }

    fn peek_next(&self) -> Option<u8> {
        self.src.get(self.pos + 1).copied()
    }

    fn bump(&mut self) -> Option<u8> {
        let ch = self.peek()?;
        self.pos += 1;
        if ch == b'\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        Some(ch)
    }

    fn skip_ws_and_comments(&mut self) {
        loop {
            while matches!(self.peek(), Some(b' ' | b'\t' | b'\r' | b'\n')) {
                self.bump();
            }

            // Comentario de línea: // ...
            if self.peek() == Some(b'/') && self.peek_next() == Some(b'/') {
                while let Some(ch) = self.bump() {
                    if ch == b'\n' {
                        break;
                    }
                }
                continue;
            }

            // Comentario de bloque: /* ... */
            if self.peek() == Some(b'/') && self.peek_next() == Some(b'*') {
                self.bump();
                self.bump();
                while let Some(ch) = self.bump() {
                    if ch == b'*' && self.peek() == Some(b'/') {
                        self.bump();
                        break;
                    }
                }
                continue;
            }

            break;
        }
    }

    fn read_ident_or_kw(&mut self) -> TokenKind {
        let start = self.pos;
        while matches!(
            self.peek(),
            Some(b'a'..=b'z' | b'A'..=b'Z' | b'_' | b'0'..=b'9')
        ) {
            self.bump();
        }
        let s = std::str::from_utf8(&self.src[start..self.pos]).unwrap();
        match s {
            "fn" => TokenKind::Fn,
            "struct" => TokenKind::Struct,
            "extern" => TokenKind::Extern,
            "import" => TokenKind::Import,
            "alias" => TokenKind::Alias,
            "return" => TokenKind::Return,
            "let" => TokenKind::Let,
            _ => TokenKind::Ident(s.to_string()),
        }
    }

    fn read_int(&mut self) -> TokenKind {
        let start = self.pos;
        while matches!(self.peek(), Some(b'0'..=b'9')) {
            self.bump();
        }
        let s = std::str::from_utf8(&self.src[start..self.pos]).unwrap();
        TokenKind::IntLit(s.parse().unwrap())
    }

    fn read_string(&mut self) -> Result<TokenKind, String> {
        self.bump(); // "
        let start = self.pos;
        while let Some(ch) = self.peek() {
            if ch == b'"' {
                let out = std::str::from_utf8(&self.src[start..self.pos])
                    .unwrap()
                    .to_string();
                self.bump();
                return Ok(TokenKind::StringLit(out));
            }
            self.bump();
        }
        Err("String sin cerrar".to_string())
    }

    pub fn tokenize(&mut self) -> Result<Vec<Token>, String> {
        let mut tokens = Vec::new();
        loop {
            self.skip_ws_and_comments();
            let line = self.line;
            let col = self.col;
            let kind = match self.peek() {
                None => TokenKind::Eof,
                Some(b'(') => {
                    self.bump();
                    TokenKind::LParen
                }
                Some(b')') => {
                    self.bump();
                    TokenKind::RParen
                }
                Some(b'{') => {
                    self.bump();
                    TokenKind::LBrace
                }
                Some(b'}') => {
                    self.bump();
                    TokenKind::RBrace
                }
                Some(b',') => {
                    self.bump();
                    TokenKind::Comma
                }
                Some(b';') => {
                    self.bump();
                    TokenKind::Semicolon
                }
                Some(b':') => {
                    self.bump();
                    TokenKind::Colon
                }
                Some(b'+') => {
                    self.bump();
                    TokenKind::Plus
                }
                Some(b'*') => {
                    self.bump();
                    TokenKind::Star
                }
                Some(b'/') => {
                    self.bump();
                    TokenKind::Slash
                }
                Some(b'=') => {
                    self.bump();
                    TokenKind::Equal
                }
                Some(b'-') => {
                    self.bump();
                    if self.peek() == Some(b'>') {
                        self.bump();
                        TokenKind::Arrow
                    } else {
                        TokenKind::Minus
                    }
                }
                Some(b'"') => self.read_string()?,
                Some(b'0'..=b'9') => self.read_int(),
                Some(b'a'..=b'z' | b'A'..=b'Z' | b'_') => self.read_ident_or_kw(),
                Some(other) => {
                    return Err(format!(
                        "Caracter no soportado '{}' en {}:{}",
                        other as char, line, col
                    ));
                }
            };
            tokens.push(Token::new(kind.clone(), line, col));
            if kind == TokenKind::Eof {
                break;
            }
        }
        Ok(tokens)
    }
}
